File: Source/Fantasy.Jobs/Sequence.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Fantasy.XSerialization;
using System.Collections;
using System.Xml.Linq;
using System.Xml;

namespace Fantasy.Jobs
{
    public abstract class Sequence : AbstractInstruction
    {
        [XArray(Serializer = typeof(InstructionsSerializer))]
        private List<IInstruction> _instructions = new List<IInstruction>();
        
        public IList<IInstruction> Instructions
        {
            get
            {
                return _instructions;
            }
        }


#pragma warning disable 169
        [XNamespace]
        protected XmlNamespaceManager _namespaces;
#pragma warning restore 169

        protected internal void ResetSequenceIndex()
        {
            IJob job = (IJob)this.Site.GetService(typeof(IJob));
            job.RuntimeStatus.Local["sequence.current"] = 0;
        }

        protected virtual void ExecuteSequence()
        {
            IJob job = (IJob)this.Site.GetService(typeof(IJob));
            int index = (int)job.RuntimeStatus.Local.GetValue("sequence.current", 0);
            while (index < this.Instructions.Count)
            {
                job.ExecuteInstruction(this.Instructions[index]);

                index++;
                job.RuntimeStatus.Local["sequence.current"] = index;
               
            }
        }

        class InstructionsSerializer : IXCollectionSerializer
        {
        #region IXCollectionSerializer Members

        public void Save(IServiceProvider context, XElement element, System.Collections.IEnumerable collection)
        {
            foreach (object inst in collection)
            {
                Type t = inst.GetType();
                XSerializer ser = new XSerializer(t) { Context = context };
                XElement childElement;
                if (t != typeof(ExecuteTaskInstruction))
                {
                    childElement = ser.Serialize(inst);
                }
                else
                {
                    ExecuteTaskInstruction taskInst = (ExecuteTaskInstruction)inst;
                    //string prefix = element.GetPrefixOfNamespace(taskInst.TaskNamespaceUri);
                    childElement = new XElement((XNamespace)taskInst.TaskNamespaceUri + taskInst.TaskName);
                    ser.Serialize(childElement, taskInst);
                }
                
                element.Add(childElement);
            }
        }

        public System.Collections.IEnumerable Load(IServiceProvider context, XElement element)
        {
            IJob job = (IJob)context.GetService(typeof(IJob));
            ArrayList rs = new ArrayList();
            foreach (XElement childElement in element.Elements())
            {
                Type t = job.ResolveInstructionType(childElement.Name);
                if (Array.IndexOf(t.GetInterfaces(), typeof(IInstruction)) < 0)
                {
                    t = typeof(ExecuteTaskInstruction);
                }
                XSerializer ser = new XSerializer(t) { Context = context };
                object inst = ser.Deserialize(childElement);
                rs.Add(inst);
            }

            return rs;
            
        }

        #endregion
    }
    }


    

}

File: branches/Stheno/Fantasy.Jobs/src/fantasy/jobs/Sequence.java
package fantasy.jobs;

import fantasy.xserialization.*;

public abstract class Sequence extends AbstractInstruction
{
//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[XArray(Serializer = typeof(InstructionsSerializer))]
	private java.util.ArrayList<IInstruction> _instructions = new java.util.ArrayList<IInstruction>();

	public final java.util.List<IInstruction> getInstructions()
	{
		return _instructions;
	}


//C# TO JAVA CONVERTER TODO TASK: There is no preprocessor in Java:
///#pragma warning disable 169
//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[XNamespace]
	protected XmlNamespaceManager _namespaces;
//C# TO JAVA CONVERTER TODO TASK: There is no preprocessor in Java:
///#pragma warning restore 169

	protected final void ResetSequenceIndex()
	{
		IJob job = (IJob)this.Site.GetService(IJob.class);
		job.getRuntimeStatus().getLocal().setItem("sequence.current", 0);
	}

	protected void ExecuteSequence()
	{
		IJob job = (IJob)this.Site.GetService(IJob.class);
		int index = (int)job.getRuntimeStatus().getLocal().GetValue("sequence.current", 0);
		while (index < this.getInstructions().size())
		{
			job.ExecuteInstruction(this.getInstructions().get(index));

			index++;
			job.getRuntimeStatus().getLocal().setItem("sequence.current", index);

		}
	}

	private static class InstructionsSerializer extends IXCollectionSerializer
	{
//C# TO JAVA CONVERTER TODO TASK: There is no preprocessor in Java:
		///#region IXCollectionSerializer Members

	public final void Save(IServiceProvider context, XElement element, Iterable collection)
	{
		for (Object inst : collection)
		{
			java.lang.Class t = inst.getClass();
			XSerializer tempVar = new XSerializer(t);
			tempVar.Context = context;
			XSerializer ser = tempVar;
			XElement childElement;
			if (t != ExecuteTaskInstruction.class)
			{
				childElement = ser.Serialize(inst);
			}
			else
			{
				ExecuteTaskInstruction taskInst = (ExecuteTaskInstruction)inst;
				//string prefix = element.GetPrefixOfNamespace(taskInst.TaskNamespaceUri);
				childElement = new XElement((XNamespace)taskInst.getTaskNamespaceUri() + taskInst.getTaskName());
				ser.Serialize(childElement, taskInst);
			}

			element.Add(childElement);
		}
	}

	public final Iterable Load(IServiceProvider context, XElement element)
	{
		IJob job = (IJob)context.GetService(IJob.class);
		java.util.ArrayList rs = new java.util.ArrayList();
		for (XElement childElement : element.Elements())
		{
			java.lang.Class t = job.ResolveInstructionType(childElement.getName());
			if (Array.indexOf(t.GetInterfaces(), IInstruction.class) < 0)
			{
				t = ExecuteTaskInstruction.class;
			}
			XSerializer tempVar = new XSerializer(t);
			tempVar.Context = context;
			XSerializer ser = tempVar;
			Object inst = ser.Deserialize(childElement);
			rs.add(inst);
		}

		return rs;

	}

//C# TO JAVA CONVERTER TODO TASK: There is no preprocessor in Java:
		///#endregion
}
}